crates/ty_ide/src/semantic_tokens.rs

```diff
@@ -0,0 +1,2531 @@
+                if let Some(module) = &import.module {
+                    // Create separate tokens for each part of a dotted module name
+                    self.add_dotted_name_tokens(module, SemanticTokenType::Namespace);
+                }
+                for alias in &import.names {
+                    if let Some(asname) = &alias.asname {
+                        // For aliased imports (from X import Y as Z), classify Z based on what Y is
+                        let ty = alias.inferred_type(self.semantic_model);
+                        let (token_type, modifiers) = self.classify_from_alias_type(ty, asname);
+                        self.add_token(asname.range(), token_type, modifiers);
+                    } else {
+                        // For direct imports (from X import Y), use semantic classification
+                        let ty = alias.inferred_type(self.semantic_model);
+                        let (token_type, modifiers) =
+                            self.classify_from_alias_type(ty, &alias.name);
+                        self.add_token(alias.name.range(), token_type, modifiers);
+                    }
+                }
+                walk_stmt(self, stmt);
+            }
+            _ => {
+                // For all other statement types, let the default visitor handle them
+                walk_stmt(self, stmt);
+            }
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &Expr) {
+        // If we have a range filter and this statement doesn't intersect, skip it
+        // as an optimization
+        if expr.range().intersect(self.range_filter).is_none() {
+            return;
+        }
+
+        match expr {
+            ast::Expr::Name(name) => {
+                let (token_type, modifiers) = self.classify_name(name);
+                self.add_token(name.range(), token_type, modifiers);
+                walk_expr(self, expr);
+            }
+            ast::Expr::Attribute(attr) => {
+                // Use semantic analysis to determine the correct token type for the attribute
+                let ty = expr.inferred_type(self.semantic_model);
+                let (token_type, modifiers) =
+                    Self::classify_from_type_for_attribute(ty, &attr.attr);
+
+                self.add_token(attr.attr.range(), token_type, modifiers);
+                // Continue visiting the base expression
+                walk_expr(self, expr);
+            }
+            ast::Expr::Call(call) => {
+                // The function being called
+                if let ast::Expr::Name(name) = call.func.as_ref() {
+                    self.add_token(name.range(), SemanticTokenType::Function, vec![]);
+                }
+                // Continue visiting arguments
+                walk_expr(self, expr);
+            }
+            ast::Expr::StringLiteral(_) => {
+                self.add_token(expr.range(), SemanticTokenType::String, vec![]);
+                walk_expr(self, expr);
+            }
```

Comment:
The `walk_expr` call seems redundant here as it'll just visit the individual string literal parts.

Is it correct to send a single token for implicitly concatenated strings or should it be separate tokens for each string literals that are present in the concatenation? For example, in `"foo" "bar"` should it be a single token for both the strings or two tokens?

If we need to have separate tokens for individual string literals, then we could use `SourceOrderVisitor::visit_string_literal` instead?